### ISSUE TITLE
[Pay 6/30] [$1000] App gets stuck in infinite loop between cash and amount page when sending money

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17637]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any report
3. Click on plus besides compose box and click on send money
4. Insert any amount and click on continue
5. Click on amount
6. Click on back button, observe it goes to new amount description page, again click on back button and observe it again opens amount page.
7. Keep on clicking back button and observe that it is stuck between new amount description page and amount page
8. Close this and now click on green plus btton
9. Insert any amount and click on continue
10. Select any user
11. Click on amount to open amount page
12. Click on back button, it will take us back to new amount description page, again click on back button and observe that now it can go all the way back to very first Send money page

## Expected Result:
App should not get stuck in infinite loop on back click and should allow user to go back to send money page like it does when we try to send money using green plus button
## Actual Result:
App gets stuck in infinite loop on back button between new amount description page and amount page
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.1.0
**Reproducible in staging?:** y
**Reproducible in production?:** new feature
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/232906963-8e2dc566-2994-41a3-a308-49951eb7f152.mp4

https://user-images.githubusercontent.com/43996225/232906980-ac4bdbc4-df05-41f5-aeb8-2b815cb0479e.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681812176434499

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010f4ffd0457045b5a</li>
        <li>Upwork Job ID: 1648783666486652928</li>
        <li>Last Price Increase: 2023-04-19</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @garrettmknight (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~010f4ffd0457045b5a

**Comment by MelvinBot** [comment]
Current assignee @garrettmknight is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @dangrous (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kmorales13** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App gets stuck in infinite loop on back button between new amount description page and amount page


### What is the root cause of that problem?
The following condition returns true regardless of the direction we are going
`const isEditingAmountAfterConfirm = currentStepIndex === 0 && previousStepIndex === _.indexOf(steps, Steps.MoneyRequestConfirm);`

This doesn't happen when using the FAB action because there's an extra `MoneyRequestParticipants` step in between.


### What changes do you think we should make in order to solve the problem?
To fix this issue we could use a history array of the steps we are navigating through instead of just the current and previous index:
`const [stepHistory, setStepHistory] = useState([0])`
This way we can add the stepIndex when we move to a next step and remove it when we go back. By having this history we can determine when we are back to the first step and stop showing the backButton.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App gets stuck in infinite loop on back button between new amount description page and amount page

### What is the root cause of that problem?
There're 2 ways to go from the `MoneyRequestConfirm` page to the `MoneyRequestAmount` page:
1. Clicking the `Amount`
2. Clicking back button in `MoneyRequestConfirm` page
Both these options will lead to the condition here https://github.com/Expensify/App/blob/c8c9974011a7e80fd96e002436a37e98bcccb2a0/src/pages/iou/MoneyRequestModal.js#L359 becomes true.
When it's true, clicking back from `MoneyRequestAmount` will always go to the `MoneyRequestConfirm` page so there's a loop.

So we can see that the condition here https://github.com/Expensify/App/blob/c8c9974011a7e80fd96e002436a37e98bcccb2a0/src/pages/iou/MoneyRequestModal.js#L359 is wrong, it should not be true when clicking back from the `MoneyRequestConfirm` page.

### What changes do you think we should make in order to solve the problem?

We need to change the way we're calculating the `isEditingAmountAfterConfirm`:
1. Introduce a new state for the `isEditingAmountAfterConfirm` instead of depending on the previous `currentStepIndex` and `previousStepIndex`.
2. In https://github.com/Expensify/App/blob/c6d4fca0c2f9b0ad4fba572f394d9b5f2bd442d1/src/pages/iou/MoneyRequestModal.js#L232, we need to check if we're going from the `MoneyRequestConfirm` page to the `MoneyRequestAmount` page. If yes then set `isEditingAmountAfterConfirm` to `true`.
3. Add an `useEffect` that listens to the `currentStepIndex`, if we go out of the `MoneyRequestAmount` page (which has index 0), we'll set `isEditingAmountAfterConfirm` back to false.

### What alternative solutions did you explore? (Optional)
There're other ways we can do the step `2` above for example by introducing a new param to `navigateToStep` and pass true in here https://github.com/Expensify/App/blob/8f5e22a56c5fc24d0c92a3916e7843361ef1922d/src/components/IOUConfirmationList.js#L328 where we're going from the `MoneyRequestConfirm` page to the `MoneyRequestAmount` page.

**Comment by garrettmknight** [comment]
@aimane-chnaif any of these proposals looking good?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mallenexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by garrettmknight** [comment]
Reassigning since I'm OOO next week.

**Comment by s77rt** [comment]
This issue seems to have the same root cause as https://github.com/Expensify/App/issues/17710 (wrong animation part) so we may handle it there. Do you want to hold / close this one?
cc @mallenexpensify @aimane-chnaif @dangrous

**Comment by s77rt** [comment]
@aimane-chnaif So do you want to handle the wrong animation here?

**Comment by dangrous** [comment]
Hey @aimane-chnaif that makes sense to me! I'll put this on hold for now

**Comment by dangrous** [comment]
Still on hold

**Comment by dangrous** [comment]
Holding still

**Comment by dangrous** [comment]
Melvin stop

**Comment by dangrous** [comment]
Still on hold, it looks like that issue might end up being on hold for another issue so this may take some time.

**Comment by mallenexpensify** [comment]
Bumped to weekly and removed `Help Wanted`. Also removed `Bug` because we're waiting on a refactor and I don't think folks assigned here aren't assigned to https://github.com/Expensify/App/issues/16526#issuecomment-1542742886 to push forward.   (ie. I don't want to be _that guy/BZ_ who's pinging that issue all the time with "any update?")

**Comment by dangrous** [comment]
Still holding

**Comment by dangrous** [comment]
still holding

**Comment by dangrous** [comment]
still on hold

**Comment by dangrous** [comment]
continues to be on hold

**Comment by dangrous** [comment]
still waiting

**Comment by dangrous** [comment]
PR fixing #17662 is on staging! We can probably move ahead with retesting and (if necessary) updating proposals

**Comment by dhanashree-sawant** [comment]
Hi @aimane-chnaif, @dangrous, the following bug was raised before 17662 so will it be eligible for reporting bonus?

**Comment by dangrous** [comment]
Yep! @dhanashree-sawant you should be paid out for the report (cc @mallenexpensify) and then I think we can close this out!

**Comment by mallenexpensify** [comment]
@dhanashree-sawant can you please accept the job and reply here once you have? https://www.upwork.com/jobs/~012afce5349d800769

**Comment by dhanashree-sawant** [comment]
Thanks @mallenexpensify, I have accepted the job.

**Comment by mallenexpensify** [comment]
@dhanashree-sawant paid! Thanks
